relayer/chains/cosmos/types.go

```diff
@@ -39,9 +50,50 @@ type packetInfo struct {
 	connectionID    string
 }
 
+// channelKey returns the processor.ChannelKey from packetInfo
+func (p packetInfo) channelKey() processor.ChannelKey {
+	return processor.ChannelKey{
+		ChannelID:             p.packet.SourceChannel,
+		PortID:                p.packet.SourcePort,
+		CounterpartyChannelID: p.packet.DestinationChannel,
+		CounterpartyPortID:    p.packet.DestinationPort,
+	}
+}
+
 // clientInfo contains the consensus height of the counterparty chain for a client.
 type clientInfo struct {
 	clientID        string
 	consensusHeight clienttypes.Height
 	header          []byte
 }
+
+type latestClientState map[string]clientInfo
+
+func (l latestClientState) UpdateLatestClientState(clientInfo clientInfo) {
+	existingClientInfo, ok := l[clientInfo.clientID]
+	if ok && clientInfo.consensusHeight.LT(existingClientInfo.consensusHeight) {
+		// height is less than latest, so no-op
+		return
+	}
+
+	// update latest if no existing state or provided consensus height is newer
+	l[clientInfo.clientID] = clientInfo
+}
+
+func (l latestClientState) Clone() latestClientState {
+	newLatestClientState := make(latestClientState)
```

Comment:
done.

> If clientInfo.header is treated as read-only

Yes, the clientInfo is used immutably. I meant to remove the `latestClientState` stuff from this PR, but the idea is that the handlers will update the latest client state with the entire clientInfo in `UpdateLatestClientState`, and a clone of the latest client state will be passed to the `PathProcessor`(s).

I removed this from this PR since it isn't applicable to the current handlers.